### DISCUSSION TITLE
Concurrent reads on the same file

### DISCUSSION THREAD

**Discussion by Nowasky** [winfsp/winfsp#463] [General]
Are concurrent reads on the same file supported by winfsp?

With this code the read method must return before a new read is executed:

```
def read(self, path, length, file_offset, fh):
    print('Executing read')
    #some code here that takes some time
    time.sleep(5)
    return 0
```

Running on Windows 10 with winfsp 1.11.22176 and fusepy 3.0.1

**Comment by billziss-gh** [comment]
Concurrent reads on the same file are supported in WinFsp. There are a few situations however that may force reads to be serialized.

**Comment by Nowasky** [comment]
Could you elaborate on what these situations would be or describe in what situations reads on the same file are not serialized? I'm trying to find a solution to this behavior.

**Comment by Nowasky** [comment]
I found a way to bypass this limitation by opening the file through another name and converting it to the original name under getattr() and read() methods. 

```
def getattr(self, path, fh=None):
    path = path.split('.zzz')[0]
    ....

def read(self, path, length, file_offset, fh):
    path = path.split('.zzz')[0]
    ....
```
So you can read a "file.txt" concurrently by opening "file.txt._sep_123", whereas .zzz is a known separator and 123 is a unique id.
